lib/Eisbach_Methods.thy

```diff
@@ -574,4 +575,156 @@ method if_concl for pattern::bool methods m =
 method case_concl for pattern::bool methods m =
   has_concl pattern, m
 
+text \<open>@{text "all"} and @{text ";"} are dangerous for weakest-precondition reasoning, as the wp
+      goals usually need to be solved in the order they were generated. The @{text fwd_all}
+      and @{text fwd_all_new} methods apply on goals in order, taking into account goals solved
+      and added.\<close>
+
+ML \<open>
+
+fun restrict_goal i n tac =
+  (PRIMITIVE (Goal.restrict i n)) THEN
+  tac THEN
+  (PRIMITIVE (Goal.unrestrict i));
```

Comment:
We should probably also put this in a `local` block or at least a struct, otherwise it's in the global name space and might get confused with the same function in the Isabelle sources.

Reply:
I think we'll need a struct, because we use it in the `method_setup`

Reply:
Good point, yes, let's use a struct then.

Reply:
I'm not sure what kind of struct to use honestly. If the only thing that does in it is restrict_goal? Should I put the other combinators there? Any bikeshed help on this one?

Reply:
I used `local` and folded the def again into the method. It's short enough.